hail/src/main/scala/is/hail/expr/ir/agg/CollectAggregator.scala

```diff
@@ -0,0 +1,102 @@
+package is.hail.expr.ir.agg
+
+import is.hail.asm4s._
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.expr.ir._
+import is.hail.expr.types.physical._
+import is.hail.utils._
+import is.hail.io.{CodecSpec, InputBuffer, OutputBuffer}
+
+/*
+ * Collect() aggregator
+ */
+
+class CollectAggregator(val elemType: PType) extends StagedAggregator {
+
+  val resultType = PArray(elemType.setRequired(false), required = true)
+
+  class State(val mb: EmitMethodBuilder) extends AggregatorState {
+    override def regionSize: Int = Region.REGULAR
+
+    /*** State ***/
+
+    val r = mb.newField[Region]
+    val region = r.load
+    val bll = new StagedBlockLinkedList(elemType, mb)
+
+    /*** Initialization ***/
+
+    val isLoaded: Code[Boolean] = region.isValid
+
+    def createState: Code[Unit] =
+      region.isNull.orEmpty(Code(
+        r := Code.newInstance[Region, Int](regionSize),
+        region.invalidate()))
+
+    def newState: Code[Unit] = Code(
+      region.getNewRegion(regionSize),
+      bll.init(region))
+
+    /*** load, store, copy ***/
+
+    val storageType = bll.storageType
+
+    def load(regionLoader: Code[Region] => Code[Unit], src: Code[Long]): Code[Unit] =
+      Code(regionLoader(region), bll.load(src))
+
+    def store(regionStorer: Code[Region] => Code[Unit], dst: Code[Long]): Code[Unit] =
+      region.isValid.orEmpty(Code(regionStorer(region), bll.store(dst), region.invalidate()))
+
+    def copyFrom(src: Code[Long]): Code[Unit] = {
+      val totalCount = region.loadInt(
+        storageType.fieldOffset(src,
+          storageType.fieldIdx("totalCount")))
+
+      totalCount.cne(0).orEmpty(
+        Code._fatal("unimplemented: copyFrom() when src is not empty"))
```

Comment:
this needs to actually create a valid state, as well---I don't see where that's happening? I think if you change this to 
```
totalCount.cne(0).mux(
  Code._fatal("unimplemented: copyFrom() when src is not empty"),
  newState)
```

Reply:
(the place this would be called is in array emitter---if you add a test (here or in python) that wraps this in an array_agg that should hit this method.)

Reply:
oh, I see what's happening. all of the initialization happens in `newState`, which is called basically anytime you expect a state to exist. I'm not sure this is wrong, per se--it seems to be doing the correct thing in all the cases, but I think I find it a little bit counterintuitive that `newState` is doing some non-trivial memory allocation.